pkg/controller/jobs/kubeflow/trainjob/trainjob_controller.go

```diff
@@ -150,10 +158,74 @@ func (t *TrainJob) PodLabelSelector() string {
 }
 
 func getChildJobSet(t *TrainJob) (*jobsetapi.JobSet, error) {
-	jobSet := jobsetapi.JobSet{}
-	// TODO: Ideally we should be using the parent context here
-	err := reconciler.client.Get(context.Background(), types.NamespacedName{Name: t.Name, Namespace: t.Namespace}, &jobSet)
-	return &jobSet, err
+	runtimeRefGK := kftrainerruntime.RuntimeRefToRuntimeRegistryKey(t.Spec.RuntimeRef)
+	runtime, ok := reconciler.runtimes[runtimeRefGK]
+	if !ok {
+		return nil, fmt.Errorf("unsupported runtime: %s", runtimeRefGK)
+	}
+
+	trainJob := (*kftrainerapi.TrainJob)(t)
+	trSpec, err := getRuntimeSpec(trainJob)
+	if err != nil {
+		return nil, fmt.Errorf("runtime '%s' not found", trainJob.Spec.RuntimeRef.Name)
+	}
+	info, err := runtime.RuntimeInfo(trainJob, trSpec.Template, trSpec.MLPolicy, trSpec.PodGroupPolicy)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get the jobsetSpecApply and apply the TrainJob object overrides for the trainer and initializer jobs
+	jobSetSpec, ok := kftrainerruntime.TemplateSpecApply[jobsetapplyapi.JobSetSpecApplyConfiguration](info)
+	if !ok {
+		return nil, err
+	}
+	jobsetApply := kftrainerjobset.NewBuilder(jobsetapplyapi.JobSet(t.Name, t.Namespace).
+		WithSpec(jobSetSpec)).Initializer(trainJob).Trainer(info, trainJob).PodLabels(info.Scheduler.PodLabels).Build()
+
+	// convert to jobset with the defaults set
+	return jobsetApplyToJobset(jobsetApply)
```

Comment:
This remains JobSet-specific so it is not ideal long term. We need to ticket and work on resolving in the future.

I guess some changes to the interfaces are still needed?

Reply:
Alternatively, it would be good if there was another helper in the Trainer which would return some struct (maybe StatusInfo) with all the details, but in Job-agnostic way.

Reply:
> Alternatively, it would be good if there was another helper in the Trainer which would return some struct (maybe StatusInfo or FullInfo) with all the details, but in Job-agnostic way.

yeah, that's the end goal, but unfortunately the info object doesn't contain all the information that ends up in the JobSet. For example, the `TrainJob.Spec.Trainer.ResourcesPerNode` property [overrides the `Trainer` `replicaJob` resources](https://github.com/kubeflow/trainer/blob/56be1f8163fb3a9b4bd999bb42e810fe429a31eb/pkg/runtime/framework/plugins/jobset/builder.go#L142) directly

Reply:
cc @andreyvelich @tenzen-y we may need to revisit the helpers and the information exposed by the Trainer so that the integration code in Kueue has all the necessary data.

@kaisoz would you like to open an issue in the Trainer repo and document what is needed to achieve fully "Job" agnostic code in Kueue?

Reply:
Ok, I opened the issue at the Kueue side: https://github.com/kubernetes-sigs/kueue/issues/7043

Reply:
Thanks @mimowo, I'll gather all the requirements and open a issue in the Trainer repo 👍🏻